src/hotspot/cpu/x86/stubGenerator_x86_64.cpp

```diff
@@ -5472,206 +5401,277 @@ address generate_avx_ghash_processBlocks() {
+      //   |   52..61    |    0..9     |     -4     |
+      //   |     62      |   + or -    | -19 or -17 |
+      //   |     63      |   / or _    | -16 or 32  |
+      //   +-------------+-------------+------------+
+      //
+      // We note that vpshufb does a parallel lookup in a destination register using
+      // the lower 4 bits of bytes from a source register.  If we use a saturated subtraction
+      // and subtract 51 from each 6-bit value, bytes from [0,51] saturate to 0, and [52,63]
+      // map to a range of [1,12].  We distinguish the [0,25] and [26,51] ranges by
+      // assigning a value of 13 for all 6-bit values less than 26.  We end up with:
+      //
+      //   +-------------+-------------+------------+
+      //   | 6-bit value |   Reduced   |   offset   |
+      //   +-------------+-------------+------------+
+      //   |    0..25    |     13      |     65     |
+      //   |   26..51    |      0      |     71     |
+      //   |   52..61    |    0..9     |     -4     |
+      //   |     62      |     11      | -19 or -17 |
+      //   |     63      |     12      | -16 or 32  |
+      //   +-------------+-------------+------------+
+      //
+      // We then use a final vpshufb to add the appropriate offset, translating the bytes.
+      // 
+      // Load input bytes - only 28 bytes.  Mask the first load to not load into the full register.
+      __ vpmaskmovd(xmm1, xmm1, Address(source, start_offset, Address::times_1, -4), Assembler::AVX_256bit);
+      __ vpshufb(xmm1, xmm1, xmm9, Assembler::AVX_256bit);
+
+      __ addl(start_offset, 24);
+
+      __ movl(rax, 0x003f03f0);
+      __ evpbroadcastd(xmm6, rax, Assembler::AVX_256bit);
+      __ movl(rax, 0x01000010);
+      __ evpbroadcastd(xmm5, rax, Assembler::AVX_256bit);
+
+      __ vpand(xmm0, xmm8, xmm1, Assembler::AVX_256bit);
+
+      __ movl(rax, 0x19191919);
+      __ evpbroadcastd(xmm3, rax, Assembler::AVX_256bit);
+      __ movl(rax, 0x33333333);
+      __ evpbroadcastd(xmm4, rax, Assembler::AVX_256bit);
+
+      __ vpmulhuw(xmm2, xmm0, xmm7, Assembler::AVX_256bit);
+      __ vpand(xmm0, xmm6, xmm1, Assembler::AVX_256bit);
+      __ vpmullw(xmm0, xmm5, xmm0, Assembler::AVX_256bit);
+      __ vpor(xmm0, xmm0, xmm2, Assembler::AVX_256bit);
+      __ vpcmpgtb(xmm2, xmm0, xmm3, Assembler::AVX_256bit);
+      __ vpsubusb(xmm1, xmm0, xmm4, Assembler::AVX_256bit);
+      __ vpsubb(xmm1, xmm1, xmm2, Assembler::AVX_256bit);
+
+      __ lea(r11, ExternalAddress(StubRoutines::x86::base64_avx2_lut_addr()));
+      __ movl(r15, isURL);
+      __ shll(r15, 5);
+      __ vmovdqu(xmm2, Address(r11, r15));
+      __ vpshufb(xmm1, xmm2, xmm1, Assembler::AVX_256bit);
+      __ vpaddb(xmm0, xmm1, xmm0, Assembler::AVX_256bit);
+
+      // Store the encoded bytes
+      __ vmovdqu(Address(dest, dp), xmm0);
+      __ addl(dp, 32);
+
+      __ cmpl(length, 31);
+      __ jcc(Assembler::belowEqual, L_process3);
+
+      __ subl(dp, 32);    // Offset for add below
```

Comment:
Is there a reason for subtracting 32 from the destination pointer and adding it back later?

Reply:
Bad logic. :-).  Fixed.